src/gui/models/scaleChargesDialogModel.cpp

```
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2024 Team Dissolve and contributors
+
+#include "gui/models/scaleChargesDialogModel.h"
+
+ScaleChargesDialogModel::ScaleChargesDialogModel() {}
+
+ScaleChargesDialogModel::~ScaleChargesDialogModel() {}
+
+double ScaleChargesDialogModel::value() const { return currentValue_; }
+
+void ScaleChargesDialogModel::updateValue(double newVal) { currentValue_ = newVal; }
+
+void ScaleChargesDialogModel::setOption(Option option) { scaleType_ = option; }
+
+ScaleChargesDialogModel::Option ScaleChargesDialogModel::getOption() { return scaleType_; }
+
+bool ScaleChargesDialogModel::scaleCharges(Species *species)
+{
+    auto scaleFactor = value();
+    if (scaleType_ != Scale)
+    {
+        auto scaleTarget = scaleFactor;
+        if (scaleTarget == 0.0)
+        {
+            return false;
+        }
```

Moving the error message dialog from `menu_species.cpp` into here, rather than returning `false` from `ScaleChargesDialogModel` to handle this specific failure, feels like a more solid and self-contained approach. This function can then just be `void`.

---

Of course, this will break your unit test, so perhaps still return the `bool` for success but add a second argument `bool showDialogOnError = true` to the `scaleCharges` function so you can prevent the dialog being raised when testing, but still test for success?

---

Yeah, I think it's best to go for one or the other, and I get the impression there is not much use returning a bool since the main work of the function is being done on the `*species` instance. In the test I will call the `void` method and then make a test on each atom to check the scaling is correct.

---

I do like the idea of the parameter to variably display the error message for testing situations